osu.Game/Storyboards/CommandTimelineGroup.cs

```diff
@@ -45,11 +45,30 @@ public CommandTimelineGroup()
             };
         }
 
+        /// <summary>
+        /// Returns the earliest visible time. Will be null unless this group has an <see cref="Alpha"/> command with a start value of zero.
+        /// </summary>
+        public double? EarliestDisplayedTime
+        {
+            get
+            {
+                var first = Alpha.Commands.FirstOrDefault();
+
+                return first?.StartValue == 0 ? first.StartTime : (double?)null;
```

Comment:
Pretty sure this check isn't needed, since a starting alpha command with other start values will still trigger an intro on stable.

In other words, if the starting alpha command starts with say, 1, this will behave as if there are no alpha-affecting commands, when there clearly is at least one.

Reply:
I also tested when a fade command has both start and end values of 0 before any other fade commands, and it still causes an intro on stable.

Reply:
To confirm, an alpha command starting at 1 makes the sprite invisible before the specified time?

Reply:
No, an alpha from 1 to 0 will make the sprite initially visible, 0 to 1 is vice versa.
If it's 1 to 1 it'll be visible during the duration of the command.
If it's 0 to 0, it'll be invisible but still affect intro time.

Reply:
I misread the question, everything else is correct but if it's 1 to 1 it'll make the sprite visible before the specified time as well. Answer is still no.

Reply:
Then the code I have is correct, I believe. Without this, *all* alpha commands will take precedence over other commands that may occur before them, which would give incorrect behaviour for the following scenario:

```
Sprite,Background,Centre,"SB\leaf.png",320,240
 S,0,500,800,0.1, 0.5
 F,0,935,1435,1,0
```

Reply:
yes it does as you expect, however, this scenario fails:

```
Sprite,Background,Centre,"SB\leaf.png",320,240
 L,-14525,1
  MX,1,0,8070,-56,645
  MY,0,0,16140,580,-100
  MX,2,8070,16140,645,-56
 S,0,935,,10
 F,0,-5000,1435,1,0.1929944
```
In stable, there is a 5 second intro, but in this branch, there is a 14 second intro (sprite is still visible and moving initially both in stable and this branch)

Reply:
Here's a clearer example. Even though the sprite visible at -14525ms because of the first fade, this still causes a 5 second intro on stable. I believe this suggests that loop groups can't affect intro time at all.

```
Sprite,Background,Centre,"SB\leaf.png",320,240
 L,-14525,1
  F,0,0,10,1,1
 S,0,935,,10
 F,0,-5000,1435,1,1
```

Reply:
Okay sure, but I think this example will still work correctly with my implementation (and going forward, be more correct?).

Is that fail case you mention actually in any beatmap? If not, that's *stable* implemented incorrectly and should be considered fine to have "fail".

Reply:
I guess any map with such patterns would simply have more of their storyboarded content visible. Alright then, seems correct.

Reply:
One thing that jumped out to me in this thread, about that

> If it's 0 to 0, it'll be invisible but still affect intro time.

case: are we talking about something like this?

```
Sprite,Background,Centre,"SB\leaf.png",320,240
 F,0,-1000,-500,0,0 // fake fade that doesn't do anything
 F,0,-500,0,0,0.5
```

In that case I'd expect the intro to start at -500 and not -1000, since the first command doesn't actually display the sprite, while it'd be -1000 using the current implementation... Is that something that storyboards rely on to start displaying earlier?

Reply:
> Is that something that storyboards rely on to start displaying earlier?

I ran a couple greps for "fades from 0 to 0" in my song folder and found nothing of the sort (among about 530 sets with storyboards). I can't remember any such cases either.

Note that the only use for this would be to extend the intro time earlier than any storyboard sprites are visible, which would only cause a black background at the start (or just the diff background if it wasn't removed in the storyboard). That was basically the exact issue with the storyboard mentioned in #11981 in lazer.

If we're going by the logic of "only visible sprites can affect intro time" as peppy seems to suggest, then yes, that case would contradict that. The question is if it needs to be fixed in this implementation, since I'd wager that nobody uses it based on my sample size (or would need/want to use it).

On another note, checking for instances of the other cases I pointed out earlier would probably take more effort than some greps. Although I could check this programmatically on all the maps I have, I don't find it worth it if we care about implementing such hypothetical storyboards 1-to-1 with stable anyway.

Reply:
It's an interesting case for sure, but I'd like to *hope* that's out of scope for the time being (until it arises that it was actually necessary for a ranked beatmap to work as expected).

One would hope that noop transforms like that would be removed by the osu! editor, or whatever scripting program the storyboarder is using.